src/v/cluster/cluster_recovery_table.cc

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/cluster_recovery_table.h"
+
+#include "cluster/cloud_metadata/cluster_manifest.h"
+#include "cluster/logger.h"
+#include "cluster/types.h"
+
+namespace cluster {
+
+bool cluster_recovery_table::validate_state_update(recovery_stage next_stage) {
+    if (_states.empty()) {
+        // Must have an existing recovery.
+        return false;
+    }
+    const auto cur_recovery = _states.back();
+    if (
+      cur_recovery.stage == recovery_stage::complete
+      || cur_recovery.stage == recovery_stage::failed) {
+        // Can't update stage after a terminal state.
+        return false;
+    }
+    switch (next_stage) {
+    case recovery_stage::initialized:
+        // Can't revert to initialized state.
+        return false;
+
+    case recovery_stage::starting:
+        // Must enter the starting stage from initialized.
+        return cur_recovery.stage == recovery_stage::initialized;
+
+    // Can reach terminal states from any non-terminal stage.
+    case recovery_stage::complete:
+    case recovery_stage::failed:
+    default:
+        // TODO: be more precise. We expect to progress linearly through
+        // stages, but it doesn't seem too problematic to allow the recovery
+        // to jump states (in the case of races).
+        return true;
+    }
+}
+
+ss::future<> cluster_recovery_table::wait_for_active_recovery() {
+    if (is_recovery_active()) {
+        co_return;
+    }
+    co_return co_await _has_active_recovery.wait();
+}
+
+std::error_code cluster_recovery_table::apply(
+  model::offset offset, cluster_recovery_init_cmd cmd) {
+    if (!_states.empty() && _states.back().is_active()) {
+        return errc::update_in_progress;
+    }
+    vlog(
+      clusterlog.info,
+      "Initializing cluster recovery at offset {} with manifest {} from bucket "
```

Comment:
what does it mean when states is non-empty, but back() state isn't active? we'd fall through to here, but does that case still correspond to initialization?

Reply:
That's right: the latest recovery attempt is complete (either it finished successfully or it failed), in which case it's okay to proceed with another, at least it won't contend with another recovery.

There's a case to be made that once successful, we should never recover another. I'm okay with making that change too, though don't feel strongly about it -- recovery is entirely an additive operation; there doesn't seem to be harm in duplicating work.